archunit/src/main/java/com/tngtech/archunit/library/modules/syntax/ModulesByAnnotationShould.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2014-2023 TNG Technology Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.library.modules.syntax;
+
+import java.lang.annotation.Annotation;
+
+import com.tngtech.archunit.PublicAPI;
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.Dependency;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.PackageMatcher;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.library.modules.AnnotationDescriptor;
+import com.tngtech.archunit.library.modules.ArchModule;
+
+import static com.tngtech.archunit.PublicAPI.Usage.ACCESS;
+
+@PublicAPI(usage = ACCESS)
+public interface ModulesByAnnotationShould<ANNOTATION extends Annotation> extends ModulesShould<AnnotationDescriptor<ANNOTATION>> {
+
+    /**
+     * Like {@link #respectTheirAllowedDependencies(DescribedPredicate, ModuleDependencyScope)}, but the allowed dependencies will be automatically
+     * derived from the {@link ANNOTATION} property named {@code annotationPropertyName}. This property *must* be of type {@code String[]}
+     * and contain the {@link ArchModule#getName() names} of the {@link ArchModule}s to which access is allowed.
+     * <br><br>
+     * For example, given the annotation
```

Comment:
> given the annotation `@MyModule`

I feel that we should emphasize more that users have to define their own annotation, and that this modules-by-annotation API is not limited to an ArchUnit-annotation (which most likely wouldn't be available for `src/main` – even though this is not a requirement; the modules could be defined in `src/test`, too).

Reply:
Maybe

```suggestion
     * For example, given the user-defined annotation
```